BUG:interpolate.Akima1DInterpolator Improving Numerical Stability

mcdigman opened the pull request:
Addresses two numerical stability concerns in the initialization of Akima Splines 

Adds two unit tests that demonstrate the fixes of the code examples in #22880 and #22882

#### Reference issue
closes #22880
closes #22882

#### What does this implement/fix?

The changes are:

1. Rearranging the equation for t in such a way that the denominator is not as susceptible to becoming enormous 
2. Handling setting the slope in the dm[1:-1]==0. case separately/explicitly 
3. Setting the finite threshold for switching formulas as the denominator of the slope approaches zero to identically 0 for method='makima', because per the citation in the documentation the method is specifically constructed to not require such a switch (see comment in #22882). 
4. Changing the scaling multiplier on the threshold to a different formula that only consider points in the same subspline. I think dm[1:-1] is a better choice than max(f12) anyway, because it implicitly appears twice in the formula for the slope, and so actually sets the scale for when the formula could potentially become numerically inaccurate.


#### Additional information

Note that in my opinion the zero-denominator threshold could also just be set to zero _always_ (i.e., also for method='akima'); the formula will be equally qualitatively discontinuous somewhere no matter where the threshold is set, and making the threshold nonzero expands the region where the formula is different to a surface with a far larger volume of points, instead of just a single point discontinuity at denom=0. Having a finite switch does not meaningfully improve the numerical stability as far as I can tell. The implementation in e.g. GSL does not have a finite cutoff like this. However, for now I left the finite cutoff present in this pull request because this is my first pull request to scipy and I do not really know whether the review standards for approving a change that (albeit slightly) changes the behavior differ from those for just a bug fix.

mcdigman committed:
BUG:interpolate.Akima1DInterpolator addresses #22880 and #22882

mcdigman commented:
Just to document the logic of the simplification/rearrangement of t I used for clarity:

t = (f1*m1+f2*m2)/(f1+f2)
  = m1 - (f1+f2)/(f1+f2)*m1 + (f1*m1+f2*m2)/(f1+f2)
  = m1 + ((-f1 - f2 + f1)*m1 + f2*m2)/(f1+f2)
  = m1 + (- f2*m1 + f2*m2)/(f1+f2)
  = m1 + (f1/(f1+f2))*(m2 - m1)

This way the division happens before the multiplication so that the multiplication involves smaller numbers, significantly mitigating the potential for overflow unless the variables were very close to overflow anyway. 

This way of writing it (with alpha = f1/(f1+f2)) is used in e.g. algorithm 13.1 in ["Akima and Renner Subsplines"](https://link.springer.com/chapter/10.1007/978-3-642-61074-5_13) from "Numerical Algorithms with C", describing the non-rounded corner akima splines method used in e.g. GSL.

j-bowhay commented:
Thanks for the contribution @mcdigman! Please be careful not to reference Numerical Recipes or GSL as they are not compatible with SciPy's license

ev-br commented:
Could you please confirm that you did not look at the GSL implementation @mcdigman ?

Like Jake said, https://github.com/scipy/scipy/pull/22888#issuecomment-2831942458, we'll need to do clean room reimplementation if you did.
Reading the docs or experimenting with input and output is OK, while reading the code itself is not, sadly.
To this end, I did not look at your implementation, and will not look at it until you confirm.

------------------------------

From reading your comments in the two issues:
- rearranging the formula to avoid a potential overflow is certainly fine;
- messing with the heuristics for the `max(f12)` threshold I'm not sure about TBH. In the end of the day, it is a heuristics, so there will always a corner case which breaks it. Making the threshold  user-controllable:  a new kwarg with the default of `1e9` instead of a hardcoded `1e9` is fine. 
- Changing the logic (threshold in the local neighborhood or globally) has backwards compatibility concerns. Given that we have a strictly local alternative in `PchipInterpolator`, I would only consider changing the logic in `Akima1DInterpolator` if there's an overwhelming evidence from the literature and alternative implementations (GSL, matlab, octave, Wolfram etc). Would be great to see a survey

mcdigman commented:
@ev-br @j-bowhay Oh no, I didn’t look at GSL’s actual code implementation or anything; I don’t know if they actually have it factored that way or not. The equation for t is just a different rearrangement from what I originally suggested in the GitHub issue that I decided was better at keeping the quantities small after playing around with the equation for a while. I didn’t mean to cite the book as a source for the equation; it is describing a totally different algorithmic modification to the original Akima splines from the ‘makima’ method, that instead uses left and right handed slopes instead of different weights like ‘makima’ does. The ‘makima’ modification is simpler and probably better in many/most applications (it lacks discontinuities). The change in factoring has nothing to do with that book; it never mentions the concept of weights at all. I can remove the mention of  it from my previous comment if it is unintentionally misleading about the origin of the equation. 

I can try to type it up more mathematically later if needed, but I think for the makima case I setting the cutoff to zero isn’t really a break in backwards compatibility so much as an improvement in the numerical behavior; I think the only case where they could disagree would be if the cutoff was misbehaving numerically. It’s a much smaller change for the makima case anyway, because it is better behaved to start with.

For the Akima case I totally agree it would be a bigger change/backwards compatibility break to set the cutoff to zero. I do not think it is quite as significant to change it from being calculated globally to locally. My concern with leaving the cutoff as calculated globally is that the failure case is totally silent, so users may be unaware that the potential for this particular misbehavior exists at all. I think it is totally plausible that someone could be interpolating say a modulated exponential function or broken power law or something and be unaware that extending the x range beyond what they strictly need could change its behavior everywhere (a factor of 10^9 is a plausible dynamic range for such a case; for example in gravitational waves we have a frequency as a function of time that can evolve all the way from the nanohertz to the kilohertz, so if I used this interpolator on that frequency function without some other mitigation, it could actually realistically trigger this issue). This could result in e.g. inconsistent behavior between different runs of the same scientific code with different boundary cutoffs (say hypothetically I evolved one run from just the nano hertz to the millihertz and was comparing to another run that was identical except that it continued all the way to the kilohertz).

ev-br commented on scipy/interpolate/_cubic.py:
```diff
@@ -535,16 +535,29 @@ def __init__(self, x, y, axis=0, *, method: Literal["akima", "makima"]="akima",
+
             # These are the mask of where the slope at breakpoint is defined:
-            ind = np.nonzero(f12 > 1e-9 * np.max(f12, initial=-np.inf))
+            ind = np.nonzero((dm[1:-1] != 0.) & (f12 > break_mult * dm[1:-1]))
```
This is potentially a backwards compat break IIUC. Some datasets will silently produce different results.

ev-br reviewed (commented):
I think I stand by my initial assessment: rearranging the formular to help numerical stability is totally fine; changing the algorithm in a backwards incompatible way is risky. Your argument in https://github.com/scipy/scipy/pull/22888#issuecomment-2832530123 make total sense --- if we were designing the function from scratch.
At this stage of the lifetime of this function, it feels that what you discuss is potential deficiencies of the algorithm. They are totally real of course. What I'm saying is I'm not convinced that fixing these potential deficiencies justifies a potential disruption of existing uses.
Now that you mention your use cases --- would Pchip work for you instead?

To summarize, I'd rather keep the `1e-9 * max(f12)` heauristics, but if, say, @j-bowhay thinks we should change it, I won't block it.

ev-br commented:
> I guess if you think this behavior should be kept in one or both 'method' cases for maximum backwards compatibility, then at least some kind of warning about this behavior could be added to the docs? Perhaps in the context of the description of a new keyword argument to set a custom threshold?

Absolutely, that'd be great!

Depending on how ambitious you feel, you may also consider adding a DIY recipe using `CubicHermiteSpline` to the tutorial (https://docs.scipy.org/doc/scipy/tutorial/interpolate.html).

The `jacobi` CI failure under free-threading is indeed unrelated and can be ignored here.

j-bowhay commented:
I think I would prefer to split this pr up. The rearrangement for numerical stability is great and could be merged straight away. The other changes might need some further discussion but shouldn't hold up the numerical stability improvements.

mcdigman committed:
added a high dynamic range test case that passes or fails specifically when m2=m3 handling is on

mcdigman commented:
@j-bowhay That makes sense; should I make this request the one for the identical improvement and then also make another, or vice versa? I haven't split a pull request in two before.

@ev-br There is one specific other case to decide how to handle, which is whether to keep my special/separate handling of the m2=m3 exactly case, where I have replaced the fill value with the analytic answer. 

As an existing comment in the code mentions, 

> if m1 == m2 != m3 == m4, the slope at the breakpoint is not 
> defined. This is the fill value:

the fill value is not really meant to be used in the m2=m3 case, because the slope _is_ defined.

This _shouldn't_ really change anything because it is an analytic simplification. However, it is possible; after some playing around, the easiest way I found to get skipping this check to trigger misbehavior is to pick a case that would _also_ trigger the high dynamic range behavior. However, I suspect it would be possible to find a case that trigger this misbehavior independently though (it might just be a less visibly obvious difference?). 

The commit I just added changes none of the functionality I previously proposed in `_cubic`, but adds a variant of the high dynamic range test case that would specifically pass or fail based on whether the m2=m3 handling was included in the old code. I attached a plot comparing the behavior it triggers; I think it is visually rougher than the other misbehavior, and really just a bug that should be fixed. 

This issue is specific to method='akima', method='makima' is almost totally unaffected. 

![akima_m2=m2_case](https://github.com/user-attachments/assets/b7a75460-74b0-4fca-91e9-d98881e3d6d0)

mcdigman commented:
I haven't yet found a case where the m2=m3 handling changes anything by more than ~floating point epsilon unless you also trigger the case where the max(f12) handling misbehaves. However I attached a plot of another case that more clearly illustrates how the behavior is worse without the m2=m3 handling in a high dynamic range case (i.e it eventually climbs to 10^10 somewhere off screen). In this case, without the m2=m3 handling it struggles to reproduce _straight lines_ even away from the discontinuity (the behavior also differs in the discontinuity, but it may be less clear which curve should be the expected behavior there).
![dynamic_range_hill_fail](https://github.com/user-attachments/assets/46cb53f2-d2c8-4f23-90f5-5ca1899385b5)

mcdigman committed:
added test that more specifically isolates the m2=m3 behavior

mcdigman committed:
made the new test case also check against the true expected result

mcdigman commented:
Oops I can fix the tests to work in 32-bit and also the linting failure. I can't make any sense of the other failures. 

I can wait to commit that until it is exactly in the form you want it to be merged for the first pull request, so let me know if you think it is ok to include the m2=m3 handling or not.

ev-br commented:
I'd suggest the following process:
- A pure maintenance PR: make a separate PR with just the identical rearrangement for numeric stability. Include a test case to show the improved stability if you have one.
- merge the maintenance PR -- this can be done today, so that it is included into the 1.16 release
- rebase on main and work on your further improvements in the 1.17 cycle.

mcdigman commented:
Sorry I missed this before the deadline. I made a new pull request rolled back to just the strictly identical changes; I left `break_mult=1e.-9` factored out as a variable (with the same hard-coded value) with the view of possibly making it a keyword argument in the future. If it is made a keyword argument, we would still need to decide whether/how to include the explicit m2=m3 case handling.

lucascolley commented:
would you like to return to this @mcdigman ?

mcdigman commented:
Ah I forgot there was more to do after the pull request merged. So, to remind myself and the thread the steps were  

- [ ]  Make the multiplier a keyword argument
- [ ] Handle m2=m3 edge case somehow. Ideas for this:
1.) Just use the analytic answer always, which more accurately reflects the intended behavior but is technically a break from backwards compatibility in a few edge cases 
2.) Use the analytic answer only if the user specifies a value for the new keyword argument, which completely prevents any break from backwards compatibility though it doesn't fix that sub-issue unless the user knows about it
3.) Use the analytic answer only if the threshold is specified and is set to exactly zero, indicating that they want all parts of this solution implemented
4.) Have a second new keyword argument to select the m2=m3 handling (IMO this would add unnecessary complexity and confusion because AFAIK there is no reason you would want the current handling other than to preserve near-identical backward compatibility)
